src/python/pants/option/subsystem.py

```diff
@@ -45,9 +75,42 @@ class Subsystem(metaclass=ABCMeta):
 
     _scope_name_re = re.compile(r"^(?:[a-z0-9_])+(?:-(?:[a-z0-9_])+)*$")
 
+    class EnvironmentAware(metaclass=ABCMeta):
+        subsystem: ClassVar[type[Subsystem]]
+        options: OptionValueContainer
+        env_tgt: EnvironmentTarget
+
+        def __getattribute__(self, __name: str) -> Any:
+            from pants.core.util_rules.environments import resolve_environment_sensitive_option
+
+            # Will raise an `AttributeError` if the attribute is not defined.
+            # MyPy should stop that from ever happening.
+            default = super().__getattribute__(__name)
+
+            # Check to see whether there's a definition of this attribute at the class level.
+            # If it returns `default` then the attribute on the instance is the same object
+            # as defined at the class, or the attribute does not exist on the class,
+            # and we don't really need to go any further.
+            v = getattr(type(self), __name, default)
+            if v is default:
+                return default
+
+            # Resolving an attribute on the class object will return the underlying descriptor.
+            # If the descriptor is an `OptionsInfo`, we can resolve it against the environment
+            # target.
+            if isinstance(v, OptionsInfo):
+                # If the the value is not defined in the `EnvironmentTarget`, return the value
+                # from the options system.
+                return resolve_environment_sensitive_option(v.flag_names[0], self) or default
+
+            # We should just return the default at this point.
+            return default
+
     @classmethod
-    def rule(cls) -> Rule:
+    def construct_subsystem_rule(cls) -> Rule:
```

Comment:
Can this be private now, in favor of calling `def rules(cls)`? Ditto `construct_env_aware_rule`.

Reply:
Yes, done!